### DISCUSSION TITLE
One file program using custom PNG images

### DISCUSSION THREAD

**Discussion by nekoalin** [Akascape/CTkMessagebox#18] [Q&A] [answered]
Hello, 

Can you make the program as one file exe using pyinstaller and custom error images? I tried it, but once I use custom images and give the location to those files, they work on the PC where the files are located, but once I move the location of the files or give the program to someone else, the pop ups won't appear (they appear in the background are are blank) 

Thank you! ^^

**Comment by Akascape** [comment, marked as answer]
@nekoalin You should change the location search method. 
I guess you are finding assets like this: `C:/some_folder/.../image.png`
This is not a good way to search files because the location may not exist in other computers.

So you should change the method for searching files.
For example, this is your folder structure:

```
‣ folder
   ↳image.png
‣ program.py
```

**Inside program.py:**
‣ simple way to search that asset:
```python
asset = "folder/image.png"
```

‣ best way to search the asset:
```python
import os
base_path = os.path.dirname(os.path.realpath(__file__))
asset = os.path.join(base_path, "folder", "image.png")
```
You should use this method to search files and images and it will never give errors.

Secondly, if you are making a one-file exe then you have to keep that folder outside the exe.

If you are packing them inside, then you have to change the search method again. That search method and a detailed guide is given here: https://github.com/TomSchimansky/CustomTkinter/discussions/939

**Reply by nekoalin** [reply]
@Akascape I will give it a try, thank you so much for the effort of teaching me this! <3 
Yes, I am new to programming and that was the method I was trying to use.

**Reply by nekoalin** [reply]
@Akascape, also, well done for the pop-ups, they look amazing!

**Reply by Akascape** [reply]
> also, well done for the pop-ups, they look amazing!

@nekoalin Thanks, just trying to contribute in the tkinter community :) and teaching the things I learned.

**Reply by nekoalin** [reply]
![image](https://github.com/Akascape/CTkMessagebox/assets/49146634/355f38ba-a6f7-4142-8943-52e22168c4af)
![image](https://github.com/Akascape/CTkMessagebox/assets/49146634/a4882961-5dbe-4e26-ad83-a354820c11ae)
Is there something wrong I did? I tried to follow your instructions. It works in the app ran from compiling, but not from standalone app made with pyinstaller

![image](https://github.com/Akascape/CTkMessagebox/assets/49146634/bce97b01-fe54-4f02-bd8a-32cf3d7b3fea)

**Reply by Akascape** [reply]
@nekoalin You will also need to keep that asset folder with the exe. (outside)

**Reply by nekoalin** [reply]
@Akascape Ohhh, so my program will require an addition folder to run properly. Hmm, okay, I understand. Is there any way to imbed that inside the program executable, as you did with default icons? They run properly without any additional folder.

**Reply by Akascape** [reply]
@nekoalin 
> Hmm, okay, I understand. Is there any way to imbed that inside the program executable, as you did with default icons? They run properly without any additional folder.

Sure you can imbed them inside. You just need to add a new search method for searching if you pack them inside.  
The full guide is given here 👉
https://github.com/TomSchimansky/CustomTkinter/discussions/939

Actually when you pack them inside, the base folder becomes different, so you have to search them inside that new location.

**Reply by Akascape** [reply]
@nekoalin 
This is what you have to do:
- First add this new function in your program.
```python
# Just add this new search function:
import sys, os
def resource(relative_path):
    base_path = getattr(
        sys,
        '_MEIPASS',
        os.path.dirname(os.path.abspath(__file__)))
    return os.path.join(base_path, relative_path)

# Then use this function to find the asset, eg: resource("my_file")
```
- use assets like this
```python
asset = resource(os.path.join('folder', 'image.png'))
```

- pack the asset folder under additional folder option of pyinstaller.

**Reply by nekoalin** [reply]
![image](https://github.com/Akascape/CTkMessagebox/assets/49146634/92e4efaa-0ca4-4555-a2e4-231af0c1843b)
@Akascape It worked! Woah! <3 
Thank you so much for the help, I couldn't do it without it, and the information you gave me is so useful and gave me so much insight about how it works. Thank you again for the time and patience! You're such a great guy!

**Reply by Akascape** [reply]
@nekoalin You can also use `cancel_button='cross'` in the ctkmessagebox in order to change the cancel button to ✖️. Just a new feature I added.

**Reply by Akascape** [reply]
@nekoalin Can I ask, how did you change the header color (title bar) of the window? Ig you are in windows 10. I generally use ctypes which works on win11.
https://github.com/TomSchimansky/CustomTkinter/discussions/1299

**Reply by nekoalin** [reply]
> @nekoalin You can also use `cancel_button='cross'` in the ctkmessagebox in order to change the cancel button to ✖️. Just a new feature I added.

ahahahah, I really wanted to ask you, actually, how to do that, but after you explained me already so much things, I didn't want to bother you even more, haha :))
I was looking at the pop-up thinking that the only thing I'd change, is that dot instead of X, but then I thought that the user is gonna press on the ok button anyway. 

I'll implement this now, thank you! ^^

**Reply by nekoalin** [reply]
> @nekoalin Can I ask, how did you change the header color of the window? Ig you are in windows 10. I generally use ctypes which works on win11. [TomSchimansky/CustomTkinter#1299](https://github.com/TomSchimansky/CustomTkinter/discussions/1299)

I'm on windows 10, didn't do anything! haha! I'm using the devices from work and we're not allowed yet to install windows 11 cause some internal tools might not be compatible. hopefully we'll get it soon, but windows 10 isn't bad, I'm ready used to it, so I'm not bothered by that.

**Reply by nekoalin** [reply]
@Akascape it looks much better! :D 
![image](https://github.com/Akascape/CTkMessagebox/assets/49146634/e0fc86ca-a3d7-4dd9-b34f-cf5bfde63f10)

As a feedback, I think it would look better if there was some type of feedback for when the mouse is hovering the X button, since the X is quite small, would help a little to make sure you're actually pressing on it.
But that's something small, overall it looks great! Thank you!

**Reply by Akascape** [reply]
> I'm on windows 10, didn't do anything! 

Lol, that was the windows theme color ig. I thought you made some changes with ctypes.

> since the X is quite small, would help a little to make sure you're actually pressing on it.

I actually disabled the mouse hover effect on that button. I don't think it will bother much. The whole area is itself a button with transparent bg.

**Reply by nekoalin** [reply]
@Akascape btw, if you have any link or something around and don't need to dig too much, I'd want to know how to to make a proper config/save system. At the beginning I had 2 frames which wrote/read from the same while, and override each other, so only one of them would get saved when I switch frames/close program. So I told them to use different config.ini files :))))) 

![image](https://github.com/Akascape/CTkMessagebox/assets/49146634/ac4f4517-9eed-4bb0-ad50-8b95718b9883)
![image](https://github.com/Akascape/CTkMessagebox/assets/49146634/0f919be2-43ca-40db-96f6-1dacc095f59f)

**Reply by Akascape** [reply]
@nekoalin Make a dictionary out of the options, save it to `.ini/.json` file types, read it again and form the dictionary and use it. I find this method easier.

**Reply by nekoalin** [reply]
@Akascape I'm still trying to process your proposal :)))) 

Correct my logic, I can create a file config.ini,  imbed it into assets folder, and use something like this(1st printscreen) to check if a value is present, and if so, to, to load config to with
 "        self.orchestration_value.set(config.get("PCMenu", "orchestration"))
        self.branch_value.set(config.get("PCMenu", "branch"))",
right? 
Then I should call this function to write into the the config file what has changed. (2nd printscreen)
But when should I call the save_config, or how, so it doesn't write over that the [App] wrote there? 


update: Hmm, and the config should be saved in such a way, so it us unpacked into a relative location when the program is opened for the first time, so if the user modifies something, it is saved somewhere, not in a temporal location like we did with the icons.png


![image](https://github.com/Akascape/CTkMessagebox/assets/49146634/def11525-1bef-4641-8ca1-f60b4506c8cc)
![image](https://github.com/Akascape/CTkMessagebox/assets/49146634/bff7f24f-361e-4b30-918f-576acff75d8c)
![image](https://github.com/Akascape/CTkMessagebox/assets/49146634/f6577b64-1eb0-4a2e-9c45-9c057834dc99)

**Comment by Akascape** [comment]
@nekoalin 
> I can create a file config.ini, imbed it into assets folder

Don't save it to the asset folder as it is packed inside the one-file exe, one-file exe treats that as a temporary folder so it will delete it after exiting the program and create a new one on every startup.
Instead, save the `.ini` file to the windows user folder.
```python
home_dir = os.getenv("UserProfile") 
save_file = os.path.join(home_dir, '.example_app', 'file.ini')
```

> But when should I call the save_config, or how, so it doesn't write over that the [App] wrote there?

save_config can be called when you exit the program. (Otherwise, you have to implement a separate button for it)
Similarly, load_config can be called when you start the program.

Here is a full example for your reference:
```python
import customtkinter
from configparser import ConfigParser
import os

def load():
    if not os.path.exists(save_file):
        return
    
    config = ConfigParser()
    config.read(save_file)
    
    # edit widgets based on the config file data
    if config.get('switch', 'toggle')=='1':
        switch.select()
        set_mode()
    textbox.insert(0.0, config.get('textbox', 'string'))
    combobox.set(config.get('combobox', 'option'))
    
def save_on_exit():
    print(f"saved file: {save_file}") # print the saved file location

    # check if base folder exists
    if not os.path.exists(os.path.join(home_dir, '.example_app')):
        os.mkdir(os.path.join(home_dir, '.example_app'))

    # set values to sections
    config = ConfigParser()
    config.add_section('switch')
    config.set('switch', 'toggle', str(switch.get()))
    config.add_section('textbox')
    config.set('textbox', 'string', textbox.get(0.0, customtkinter.END))
    config.add_section('combobox')
    config.set('combobox', 'option', combobox.get())

    # save the file 
    with open(save_file, 'w') as f:
        config.write(f)
        
    app.destroy()
    
def set_mode():
    if switch.get()==1:        
        customtkinter.set_appearance_mode("Dark")
    else:
        customtkinter.set_appearance_mode("Light")

# save the file in the user folder    
home_dir = os.getenv("UserProfile") 
save_file = os.path.join(home_dir, '.example_app', 'file.ini')
            
app = customtkinter.CTk()
app.title("Example App")
app.geometry("250x300")

# save file on exit
app.protocol("WM_DELETE_WINDOW", save_on_exit)

# some widgets
frame_1 = customtkinter.CTkFrame(master=app)
frame_1.pack(expand=True, fill="both")

switch = customtkinter.CTkSwitch(master=frame_1, text="Dark Mode", command=set_mode)
switch.pack(pady=10)
set_mode()

textbox = customtkinter.CTkTextbox(master=frame_1)
textbox.pack()

combobox = customtkinter.CTkComboBox(master=frame_1, values=["A", "B", "C"])
combobox.pack(pady=10)

load() # load the data

app.mainloop()                
```

**Reply by nekoalin** [reply]
@Akascape, thank you a lot, I will try to digest the information, then apply it to my code. 

What's a little confusing for me is that I have save system in multiple app files (I have window for PC, one for Xbox, both of them have different widgets), and here comes my confusion of how to implement it. If I apply your save system onto 2 different files, won't both of them try to override each other?

**Reply by Akascape** [reply]
@nekoalin 
In that case you have to change the file names for both the app as they have different versions.
For a complex file save system, try to implement a save/load button. Use the tkinter `asksaveasfilename` method to save the file according to the user. The user can then load that specific file.

**Reply by nekoalin** [reply]
@Akascape thank you so much for the explanations! I really appreciate that a lot! I decided that for now, till I get better at programming and my my logic can understand it, will go with multiple config.inis for each frame, but will take your approach and save them inside user's Documents, so if the app is on desktop, the config files won't be there. 
I don't think it's a problem if there are multiple config files as long as the user is not bothered by them.
